# Chapter 4 Exercises
# Algorithm Workbench

## Problem 1
Write a while loop that lets the user enter a number. The number should be multiplied by 10, and the result assigned to a variable named product. The loop should iterate as long as product is less than 100.

In [5]:
# Get number from user
number = float(input('Enter a number: '))

# Formula to calculate product 
product = number * 10

# While loop that executes calculation while product is less than 100
while product < 100:
    product = number * 10
    print(product)
    number = float(input('Enter a number: '))

Enter a number: 9.9
99.0
Enter a number: 9.99
99.9
Enter a number: 9.123456789
91.23456789000001
Enter a number: 10
100.0
Enter a number: 10.01


## Problem 2 
Write a while loop that asks the user to enter two numbers. The numbers should be added and the sum displayed. The loop should ask the user if he or she wishes to perform the operation again. If so, the loop should repeat, otherwise it should terminate.

In [ ]:
# Get two numbers from user
number1 = int(input('Enter a number: '))
number2 = int(input('Enter a number: '))

total = number1 + number2

# Establish continuation of while loop
keep_going = 'y'

# While loop to add two numbers, display sum, and request input
# for another calculation
while keep_going == 'y' or keep_going == 'Y':
    print(total)
    input('Do you want to add another set of numbers? ')
    number1 = int(input('Enter a number: '))
    number2 = int(input('Enter a number: '))

Enter a number: 4
Enter a number: 4
8
Do you want to add another set of numbers? y
Enter a number: 7
Enter a number: 3
8
Do you want to add another set of numbers? n


I noticed I didn't have the variable "total" inside of the while loop or keep_going assigned to the input within the loop

In [ ]:
# Get two numbers from user
number1 = int(input('Enter a number: '))
number2 = int(input('Enter a number: '))

# Establish continuation of while loop
keep_going = 'y'

# While loop to add two numbers, display sum, and request input 
# for another calculation
while keep_going == 'y' or keep_going == 'Y':
    total = number1 + number2
    print(total)
    keep_going = input('Do you want to add another set of numbers? ')
    number1 = int(input('Enter a number: '))
    number2 = int(input('Enter a number: '))

## Problem 3
Write a for loop that displays the following set of numbers:

 0, 10, 20, 30, 40, 50 . . . 1000

In [3]:
# For loop that displays numbers 0 through 1000
# in factors of 10
for i in range(0, 1001, 10):
    print(i)

0
10
20
30
40
50
60
70
80
90
100
110
120
130
140
150
160
170
180
190
200
210
220
230
240
250
260
270
280
290
300
310
320
330
340
350
360
370
380
390
400
410
420
430
440
450
460
470
480
490
500
510
520
530
540
550
560
570
580
590
600
610
620
630
640
650
660
670
680
690
700
710
720
730
740
750
760
770
780
790
800
810
820
830
840
850
860
870
880
890
900
910
920
930
940
950
960
970
980
990
1000


## Problem 4
Write a loop that asks the user to enter a number. The loop should iterate 10 times and keep a running total of the numbers entered.

In [107]:
# Get number from user
number = int(input('Enter a number: '))

# Establish constant that is starting point of range
START = 0

# Establish constant that is ending point of range
FINISH = 10

# Running total variable 
total = 0

# for loop that requests user to input a number
# program will display the number and
# keep a running total of numbers entered
for i in range(START, FINISH):
    total += 1
    print(f'Number: {number}\n Total: {total + 1}')
    number = int(input('Enter a number: '))

Enter a number: 45
Number: 45
 Total: 2
Enter a number: 45
Number: 45
 Total: 3
Enter a number: 6
Number: 6
 Total: 4
Enter a number: 7
Number: 7
 Total: 5
Enter a number: 4
Number: 4
 Total: 6
Enter a number: 4657
Number: 4657
 Total: 7
Enter a number: 45
Number: 45
 Total: 8
Enter a number: 34
Number: 34
 Total: 9
Enter a number: 46
Number: 46
 Total: 10
Enter a number: 67
Number: 67
 Total: 11
Enter a number: 456


#### With the above program, I could not figure out how to display the final total statement without having to enter 11 numbers 

# Programming Exercises
## Problem 1 - Bug Collector
A bug collector collects bugs every day for five days. Write a program that keeps a running total of the number of bugs collected during the five days. The loop should ask for the number of bugs collected for each day, and when the loop is finished, the program should display the total number of bugs collected.

In [105]:
# Constant of 5 days per week
DAYS = 5

# Running total accumulator 
total = 0

# For loop that iterates each day of DAYS
# and requests input for bugs collected each day
# Running total is total bugs per week 
for day_num in range(DAYS):
    bugs = int(input(f'Enter the number of bugs collected during day {day_num + 1} of the week: '))
    total += bugs

# Display total number of bugs collected
print(f'The total number of bugs collected this week is: {total}')

Enter the number of bugs collected during day 1 of the week: 5
Enter the number of bugs collected during day 2 of the week: 5
Enter the number of bugs collected during day 3 of the week: 5
Enter the number of bugs collected during day 4 of the week: 5
Enter the number of bugs collected during day 5 of the week: 5
The total number of bugs collected this week is: 25


## Problem 4 - Distance Travelled
The distance a vehicle travels can be calculated as follows:

distance = speed x time

For example, if a train travels 40 miles per hour for three hours, the distance traveled is 120 miles. Write a program that asks the user for the speed of a vehicle (in miles per hour) and the number of hours it has traveled. It should then use a loop to display the distance the vehicle has traveled for each hour of that time period. 

In [17]:
# Get speed and hours travelled from user
speed = int(input('Enter the speed travelled in mph: '))
hours_travelled = int(input('Enter the amount of hours travelled: '))

# Running total for total distance travelled
total_distance = 0

# Input validation loop that does 
# not allow negative numbers
while speed < 0 or hours_travelled < 0:
    print('Error: Invalid data entry. Please enter a valid\n'
         + 'number in the \"speed\" or \"hours travelled\" section.')
    speed = int(input('Enter the speed travelled at in mph: '))
    hours_travelled = int(input('Enter the amount of hours travelled: '))

# for loop that calculates miles travelled 
# at each hour of travel 
for hr in range(hours_travelled):
    distance_in_hr = speed * 1
    total_distance += distance_in_hr
    print(total_distance)

Enter the speed travelled in mph: -10
Enter the amount of hours travelled: 55
Error: Invalid data entry. Please enter a valid
number in the "speed" or "hours travelled" section.
Enter the speed travelled at in mph: 70
Enter the amount of hours travelled: 10
70
140
210
280
350
420
490
560
630
700


## Problem 11
If a moderately active person cuts their calorie intake by 500 calories a day, they can typically lose about 4 pounds a month. Write a program that lets the user enter their starting weight, then creates and displays a table showing what their expected weight will be at the end of each month for the next 6 months if they stay on this diet.

In [72]:
# Get starting weight from user
starting_weight = int(input('Enter your starting weight: '))

# Constant for rate of loss 4 lb/month
LOSS_RATE = 4

# Constant for length of diet in months
DIET_LENGTH = 6

# Running total of starting weight to be subtracted from 
expected_weight = starting_weight

# For loop that calculates weight at each
# month during the diet and prints result
# in table format 
for month in range(DIET_LENGTH):
    monthly_weight = expected_weight - LOSS_RATE
    expected_weight -= LOSS_RATE
    print(f'|Expected weight at end of month {month + 1}|')
    print(f'|\t\t{monthly_weight}\t\t  |')
    print('|---------------------------------|')
    print('|\t\t\t\t  |')

Enter your starting weight: 200
|Expected weight at end of month 1|
|		196		  |
|---------------------------------|
|				  |
|Expected weight at end of month 2|
|		192		  |
|---------------------------------|
|				  |
|Expected weight at end of month 3|
|		188		  |
|---------------------------------|
|				  |
|Expected weight at end of month 4|
|		184		  |
|---------------------------------|
|				  |
|Expected weight at end of month 5|
|		180		  |
|---------------------------------|
|				  |
|Expected weight at end of month 6|
|		176		  |
|---------------------------------|
|				  |


The most difficult part for this one was formatting the table. I ended up centering the monthly weight using \t, although I'm sure there is a better way to do it. Also couldn't find out how not to print the last portion of the table again. 

# Problem 14
Write a program that uses nested loops to draw this pattern:
 *******
 ******
 *****
 ****
 ***
 **
 *

In [103]:
# Set base size for triangle since it is 7 by 7
BASE_SIZE = 7

# for loop that prints inverted triangle of asterisks
for row in range(BASE_SIZE):
    for col in range(BASE_SIZE - row):
        print('*', end='')
    print()

*******
******
*****
****
***
**
*


This one took more time than I expected. I found that if you reverse the loop in the inner for loop, it will invert the triangle from the original code we had from the book. 

## Problem 10 Tuition increase
At one college, the tuition for a full-time student is $8,000 per semester. It has been announced that the tuition will increase by 3 percent each year for the next 5 years. Write a program with a loop that displays the projected semester tuition amount for the next 5 years.

In [157]:
# Constant rate of 3% increase over 5 years
INCREASE_PER_YEAR = 0.03

YEARS = 5 

# Running total to keep track of total
# cost per year after increase
final_tuition_cost = 8000

# Table header
print('TUITION INCREASES OVER THE NEXT 5 YEARS')
print('---------------------------------------')

# for loop calculates cost after increase each year
# displays result
for year in range(YEARS):
    final_tuition_cost += (INCREASE_PER_YEAR * final_tuition_cost) * 3
    print(f'Tutition cost after year {year + 1} of increase:\n', '{:,.2f}'.format(final_tuition_cost), '\n')

TUITION INCREASES OVER THE NEXT 5 YEARS
---------------------------------------
Tutition cost after year 1 of increase:
 8,720.00 

Tutition cost after year 2 of increase:
 9,504.80 

Tutition cost after year 3 of increase:
 10,360.23 

Tutition cost after year 4 of increase:
 11,292.65 

Tutition cost after year 5 of increase:
 12,308.99 

